File: src/main/scala/com/wikibooks/spark/ch7/scala/ForeachSample.scala
package com.wikibooks.spark.ch7.scala

import org.apache.spark.sql.{ForeachWriter, Row, SparkSession}
import org.apache.spark.sql.streaming.OutputMode

// 7.4.4.4
object ForeachSample {

  def main(args: Array[String]): Unit = {

    import org.apache.spark.sql.functions._

    val spark = SparkSession
      .builder()
      .appName("ForeachSample")
      .master("local[*]")
      .getOrCreate()

    import spark.implicits._

    // port번호는 ncat 서버와 동일하게 수정해야 함
    val lines = spark
      .readStream
      .format("socket")
      .option("host", "localhost")
      .option("port", 9000)
      .load()

    // step4
    val words = lines.select(explode(split('value, " ")).as("word"))
    val wordCount = words.groupBy("word").count

    // step5(foreach 메서드는 스칼라 및 자바만 사용 가능)
    val query = wordCount
      .writeStream
      .outputMode(OutputMode.Complete)
      .foreach(new ForeachWriter[Row] {
        def open(partitionId: Long, version: Long): Boolean = {
          println(s"partitionId:${partitionId}, version:${version}")
          true
        }

        def process(record: Row) = {
          println("process:" + record.mkString(", "))
        }

        def close(errorOrNull: Throwable): Unit = {
          println("close")
        }
      }).start()

    // step6
    query.awaitTermination()
  }
}

File: src/main/java/com/wikibooks/spark/ch7/ForeachSample.java
package com.wikibooks.spark.ch7;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.ForeachWriter;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.streaming.OutputMode;
import org.apache.spark.sql.streaming.StreamingQuery;

import static org.apache.spark.sql.functions.*;

public class ForeachSample {

  // 7.2절
  public static void main(String[] args) throws Exception {

    SparkSession spark = SparkSession
            .builder()
            .appName("ForeachSample")
            .master("local[*]")
            .getOrCreate();

    // port번호는 ncat 서버와 동일하게 수정해야 함
    Dataset<Row> lines = spark
            .readStream()
            .format("socket")
            .option("host", "localhost")
            .option("port", 9000)
            .load();

    Dataset<Row> words = lines.select(explode(split(col("value"), " ")).as("word"));
    Dataset<Row> wordCount = words.groupBy("word").count();

    // foreach 메서드는 스칼라 및 자바만 사용 가능
    StreamingQuery query = wordCount
            .writeStream()
            .outputMode(OutputMode.Complete())
            .foreach(new ForeachWriter<Row>() {
              @Override
              public boolean open(long partitionId, long version) {
                System.out.println("partitionId:" + partitionId + "version:" + version);
                return true;
              }

              @Override
              public void process(Row value) {
                System.out.println("process:" + value.mkString(", "));
              }

              @Override
              public void close(Throwable errorOrNull) {
                System.out.println("close");
              }
            }).start();

    query.awaitTermination();
  }
}
